### DISCUSSION TITLE
How can I get the ansi of a string?

### DISCUSSION THREAD

**Discussion by otakutyrant** [nushell/nushell#10090] [Q&A] [answered]
I want to create a powerline prompt so I need to get the current bg ansi and fg ansi of a string to use them to render the following triangle symbol accordingly.

**Comment by fdncred** [comment, marked as answer]
You should have a look at nu_scripts/modules/prompts/oh-my.nu. It looks like this.
![image](https://github.com/nushell/nushell/assets/343840/54b61ee0-1a7c-4c50-a89b-0554cfa3415a)

To answer your question, unless you have the code, I'm not sure there's a way to detect the ansi color of text without reading the ansi codes.

You can do this with ansi escape sequences that color the text you're creating. You can see here that `red` is mapped to `esc[31m`, then there's the word `foo`, then reset with `esc[0m`.
```nushell
❯ $"(ansi red)foo(ansi reset)" | debug -r
String {
    val: "\u{1b}[31mfoo\u{1b}[0m",
    span: Span {
        start: 327939,
        end: 327967,
    },
}
```

**Reply by amtoine** [reply]
i wish we could do 
```nushell
> $"(ansi red)foo(ansi reset)" | debug -r | get val
\u{1b}[31mfoo\u{1b}[0m
```
:eyes:

**Reply by fdncred** [reply]
You can do this.
```nushell
❯ $"(ansi red)foo(ansi reset)" | debug -r | parse -r 'val: "(?P<ansi>.*)",'
╭#┬─────────ansi─────────╮
│0│\u{1b}[31mfoo\u{1b}[0m│
╰─┴──────────────────────╯
```
or this
```nushell
❯ $"(ansi red)foo(ansi reset)" | debug -r | parse -r 'val: "(?P<ansi>.*)",' | get ansi.0
\u{1b}[31mfoo\u{1b}[0m
```

**Reply by amtoine** [reply]
true but ehe :smirk:

**Comment by panicbit** [comment]
You are all thinking too hard about it apparently :sweat_smile:

```
❯ $"(ansi red)foo(ansi reset)" | str replace -a '\e' '\e'
\e[31mfoo\e[0m
```

PS: don't forget the `-a` :sweat_smile:

**Reply by fdncred** [reply]
Replacing `'\e'` with `'\e'` doesn't do anything that i can tell.

**Reply by fdncred** [reply]
But if you add double quotes to the first one, it works, and is indeed easier. Thanks for the tip @panicbit!
```nushell
$"(ansi red)foo(ansi reset)" | str replace -a "\e" '\e'
\e[31mfoo\e[0m
```

**Reply by panicbit** [reply]
If it's not working with the single quotes for you, then we have to look into it, because it clearly does work for me.

**Reply by fdncred** [reply]
It doesn't work for me in Windows, nor in Ubuntu. Double quotes works in nushell because `"\e"` is treated as the `escape` character aka `\u{1b}`, whereas `'\e'` is treated as a string in nushell. It's odd that this would work for you. What OS are you on? I haven't tried MacOS yet.

**Reply by panicbit** [reply]
Oh wow, I thought that `\e` would be accepted by re2 / the regex crate, but [that is clearly not the case](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7db3433b3a5fc2866081423ea0497803).
Looks like I am the odd one out here :sweat_smile: 

| key                | value                                                                           |
| ------------------ | ------------------------------------------------------------------------------- |
| version            | 0.83.1                                                                          |
| branch             |                                                                                 |
| commit_hash        |                                                                                 |
| build_os           | linux-x86_64                                                                    |
| build_target       | x86_64-unknown-linux-gnu                                                        |
| rust_version       | rustc 1.71.0 (8ede3aae2 2023-07-12)                                             |
| rust_channel       | stable-x86_64-unknown-linux-gnu                                                 |
| cargo_version      | cargo 1.71.0 (cfd3bbd8f 2023-06-08)                                             |
| build_time         | 2023-08-02 12:40:14 +02:00                                                      |
| build_rust_channel | release                                                                         |
| allocator          | mimalloc                                                                        |
| features           | dataframe, default, extra, sqlite, static-link-openssl, trash, wasi, which, zip |
| installed_plugins  |                                                                                 |

**Reply by fdncred** [reply]
Oh, I bet I know what it is. In `0.83.1` `str replace` is defaulting to using regex. In `0.83.2`, which i'm using, it defaults to using `string`. So, if I explicitly tell it to use regex with `-r`, it works like yours `$"(ansi red)foo(ansi reset)" | str replace -ar '\e' '\e'`.

**Reply by panicbit** [reply]
I think it's working for you because of https://github.com/nushell/nushell/pull/10038.
Try with `-r` (enabling the regex mode again):

```
$"(ansi red)foo(ansi reset)" | str replace -r -a '\e' '\e'
```
This narrows the quirk down to the regex codepath :smile:. I'll investigate further.

**Reply by fdncred** [reply]
Yup, we were typing the same thing at the same time. That's what is going on.

**Reply by panicbit** [reply]
So, it turns out that nushell actually uses [fancy-regex](https://docs.rs/fancy-regex/) here and that crate apparently accepts `\e` as escape :sweat_smile:.

**Reply by panicbit** [reply]
![image](https://github.com/nushell/nushell/assets/628445/1c84f319-be17-42b6-915e-04fa9f7973c0)

**Reply by fdncred** [reply]
Right, but only as long as you're using regex, which is a change in the upcoming release today.
